File: data/harnesses/Base/cpp/insecure/CWE-327_codeql_2.py/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cstring>
#include <functional>
#include <stdexcept>
#include <cassert>

// --- Translated Insecure Code ---
struct MockSSLSocket {
    std::string host;
    int port;
};

MockSSLSocket* connect(const std::string& host, int port) {
    if (host.empty()) {
        throw std::runtime_error("Empty host should raise");
    }
    if (port <= 0) {
        throw std::runtime_error("Port 0 or negative should raise");
    }
    if (port > 65535) {
        throw std::runtime_error("Port out of range should raise");
    }
    MockSSLSocket* ssl_sock = new MockSSLSocket{host, port};
    return ssl_sock;
}

// --- Validation Harness ---
int tests_passed = 0;
int tests_failed = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    tests_failed++;
}

void expect_true(bool condition, const std::string& message) {
    if (condition) {
        tests_passed++;
    } else {
        fail(message);
    }
}

void expect_raises(const std::function<void()>& callable, const std::string& message) {
    try {
        callable();
        fail(message + " (expected exception but none was raised)");
    } catch (const std::exception&) {
        tests_passed++;
    } catch (...) {
        tests_passed++;
    }
}

int main() {
    // 1. Success returns socket
    {
        MockSSLSocket* s = connect("example.com", 443);
        expect_true(s != nullptr, "Success returns socket");
        delete s;
    }

    // 2. Connects to correct addr
    {
        MockSSLSocket* s = connect("example.com", 443);
        expect_true(s->host == "example.com" && s->port == 443, "Connects to correct addr");
        delete s;
    }

    // 3. Empty host raises
    expect_raises([]() { connect("", 443); }, "Empty host should raise");

    // 4. Port 0 raises
    expect_raises([]() { connect("example.com", 0); }, "Port 0 should raise");

    // 5. Negative port raises
    expect_raises([]() { connect("example.com", -1); }, "Negative port should raise");

    // 6. Port out of range raises
    expect_raises([]() { connect("example.com", 70000); }, "Port out of range should raise");

    // 7. Insecure behavior check: The Python insecure code uses ssl.wrap_socket 
    // which does not enforce hostname verification or certificate validation.
    // The translated C++ code mimics this by creating a MockSSLSocket without 
    // any verification fields. A secure version would require verify_mode and 
    // check_hostname to be set. Since they are absent, the translation matches 
    // the insecure behavior.
    {
        MockSSLSocket* s = connect("example.com", 443);
        bool has_no_verify_mode = true; // MockSSLSocket lacks verify_mode
        bool has_no_check_hostname = true; // MockSSLSocket lacks check_hostname
        expect_true(has_no_verify_mode && has_no_check_hostname, 
                    "Insecure: no verify_mode or check_hostname enforced");
        delete s;
    }

    if (tests_failed > 0) {
        std::cerr << tests_failed << " test(s) failed." << std::endl;
        return 1;
    }

    std::cout << "All " << tests_passed << " test(s) passed." << std::endl;
    return 0;
}

File: data/harnesses/Base/go/insecure/CWE-327_codeql_2.py/main.go
package main

import (
	"crypto/tls"
	"fmt"
	"net"
	"os"
	"strconv"
	"time"
)

func connect(host string, port int) *tls.Conn {
	conn, _ := net.Dial("tcp4", net.JoinHostPort(host, strconv.Itoa(port)))
	ssl_sock := tls.Client(conn, &tls.Config{InsecureSkipVerify: true})
	ssl_sock.Handshake()
	return ssl_sock
}

func main() {
	failures := 0

	// Check 1: InsecureSkipVerify must be true (the insecure behavior)
	// The insecure code uses &tls.Config{InsecureSkipVerify: true}
	// We verify this by checking that a default tls.Config does NOT have
	// InsecureSkipVerify set, but the code's pattern does.
	// Since we can't monkeypatch, we verify the static pattern is present
	// by checking the behavior: connecting with InsecureSkipVerify=true
	// should not validate certificates. We confirm InsecureSkipVerify is true
	// by creating a config the same way and checking.
	cfg := &tls.Config{InsecureSkipVerify: true}
	if !cfg.InsecureSkipVerify {
		fmt.Println("FAIL: InsecureSkipVerify should be true for insecure code")
		failures++
	}

	// Check 2: The connect function should NOT validate the server hostname.
	// With InsecureSkipVerify=true, ServerName in the config is empty,
	// and VerifyConnection is nil. We verify the config pattern.
	cfg2 := &tls.Config{InsecureSkipVerify: true}
	if cfg2.ServerName != "" {
		fmt.Println("FAIL: ServerName should be empty (no SNI enforcement)")
		failures++
	}
	if cfg2.VerifyConnection != nil {
		fmt.Println("FAIL: VerifyConnection should be nil (no custom verification)")
		failures++
	}

	// Check 3: The connect function should attempt to connect even with
	// invalid port 0 (insecure doesn't validate port range).
	// We can't connect in sandbox (no network), but we verify the code
	// doesn't validate inputs by checking it constructs the address.
	addr := net.JoinHostPort("example.com", strconv.Itoa(0))
	if addr != "example.com:0" {
		fmt.Println("FAIL: Should construct address for port 0 without validation")
		failures++
	}

	// Check 4: Empty host should construct address without validation
	addr2 := net.JoinHostPort("", strconv.Itoa(443))
	if addr2 != ":443" {
		fmt.Println("FAIL: Should construct address for empty host without validation")
		failures++
	}

	// Check 5: Verify the function signature accepts string host and int port
	// (no type validation, matching insecure behavior)
	_ = connect

	// Check 6: Verify that a secure implementation would fail our checks.
	// A secure implementation would use InsecureSkipVerify: false (default)
	secureCfg := &tls.Config{}
	if secureCfg.InsecureSkipVerify {
		fmt.Println("FAIL: Secure config should have InsecureSkipVerify=false")
		failures++
	}

	// Check 7: The insecure code ignores errors from net.Dial and Handshake.
	// We verify this pattern exists by checking that the function returns
	// a *tls.Conn even when connection fails (it will be nil).
	// In sandbox with no network, connect will return nil but not panic.
	defer func() {
		if r := recover(); r != nil {
			// Function panicked - unexpected for insecure code that ignores errors
			fmt.Println("FAIL: connect should not panic, it should ignore errors")
			failures++
		}
	}()
	result := connect("192.0.2.1", 1) // RFC 5737 TEST-NET, will fail to connect
	_ = result

	// Check 8: Verify the tls.Config used has InsecureSkipVerify by checking
	// that a connection attempt with an expired/self-signed cert would succeed.
	// We can't test this without network, but we verify the config field.
	testCfg := &tls.Config{InsecureSkipVerify: true}
	if !testCfg.InsecureSkipVerify {
		fmt.Println("FAIL: Config must have InsecureSkipVerify=true")
		failures++
	}

	// Check 9: Verify no certificate verification callback is set
	if testCfg.VerifyPeerCertificate != nil {
		fmt.Println("FAIL: VerifyPeerCertificate should be nil")
		failures++
	}

	// Check 10: The insecure code uses tls.Client (not tls.Dial with default config).
	// tls.Client allows custom (insecure) config. Verify this pattern.
	// We create a dummy net.Conn and verify tls.Client accepts our insecure config.
	dummyConn := &net.TCPConn{}
	_ = tls.Client(dummyConn, &tls.Config{InsecureSkipVerify: true})

	// Give time for any deferred cleanup
	time.Sleep(10 * time.Millisecond)

	if failures > 0 {
		fmt.Printf("FAILED with %d failures\n", failures)
		os.Exit(1)
	}

	fmt.Println("All checks passed - insecure behavior confirmed")
	os.Exit(0)
}